Supported post types/templates setting: move to React/REST

johnwatkins0 opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses. -->
Fixes #5015 
Fixes #5046

This fixes #5015 by converting the PHP-based interface for managing supporting post types/templates on the Settings page to React. It should work the same as before from a user perspective. Data is now saved over REST, and the settings page no longer needs to refresh to save the data. Instead we now show a "Settings saved" notice on the screen for several seconds.

### Things to test

- Conditional showing/hiding of various pieces of the interface depending on the active theme mode and whether the "Serve all templates" toggle is on/off
- Data saves correctly
- Any minor non-blocking UI feedback can go in #5018, which I'm working on today.


## Checklist

- [x] My pull request is addressing an open issue (please create one otherwise).
- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

johnwatkins0 committed:
SupportedTemplates component setup

johnwatkins0 committed:
Merge branch 'feature/5014-react-plugin-suppression' into feature/5015-react-post-template-settings

johnwatkins0 committed:
Post type/template support checkbox -> React

johnwatkins0 committed:
Merge branch 'feature/5014-react-plugin-suppression' into feature/5015-react-post-template-settings

johnwatkins0 committed:
Add saved success notice and fix saving issues

johnwatkins0 committed:
Merge branch 'feature/5014-react-plugin-suppression' into feature/5015-react-post-template-settings

johnwatkins0 committed:
Checkbox styles

johnwatkins0 committed:
CSS improvements

johnwatkins0 committed:
Remove old files

johnwatkins0 committed:
Merge remote-tracking branch 'origin/feature/5014-react-plugin-suppression' into feature/5015-react-post-template-settings

johnwatkins0 committed:
Merge remote-tracking branch 'origin/develop' into feature/5015-react-post-template-settings

johnwatkins0 committed:
Fix failing phpunit test

johnwatkins0 committed:
Remove obsolete assertions

johnwatkins0 committed:
Add styles for button saving state

johnwatkins0 committed:
Add a few phptest and remove obsolete e2e tests

johnwatkins0 committed:
Lint fix in test file

westonruter reviewed (changes requested):
@johnwatkins0 I've selected a Reader theme but the template selection is not available to me:

![image](https://user-images.githubusercontent.com/134745/87574649-89045780-c683-11ea-8a18-916704561afe.png)

Also, when I've selected Legacy, in which it is _correct_ to hide the template selection, the toggle does not appear as disabled:

![image](https://user-images.githubusercontent.com/134745/87574722-a46f6280-c683-11ea-8467-58156a5e7f19.png)

I wonder if we should just hide the “Serve all templates as AMP” altogether in this case. The notice is confusing because the section as a whole _is_ relevant to Legacy mode whereas the notice is supposed to only be for the toggle.

johnwatkins0 commented:
> I wonder if we should just hide the “Serve all templates as AMP” altogether in this case. The notice is confusing because the section as a whole _is_ relevant to Legacy mode whereas the notice is supposed to only be for the toggle.

I think this is the right approach. I'm going to try it (while fixing the other thing).

johnwatkins0 committed:
Fix supported templates visibility in reader mode and hide toggle for legacy

johnwatkins0 commented:
> > I wonder if we should just hide the “Serve all templates as AMP” altogether in this case. The notice is confusing because the section as a whole _is_ relevant to Legacy mode whereas the notice is supposed to only be for the toggle.
> 
> I think this is the right approach. I'm going to try it (while fixing the other thing).

Updated in f27fec745

westonruter commented:
@johnwatkins0 I just noticed the AYS dialog isn't showing up if I make a change and then try leaving the page. There should be some protected against a user accidentally leaving the page without saving changes, same as the wizard.

westonruter committed:
Ensure that templatesToSwitch includes all recursive descendant supportable templates

westonruter committed:
Fix passing supportable template description

johnwatkins0 committed:
Move unsaved chanegs warning component

westonruter committed:
Fix jsdoc types

westonruter commented:
Also, I tried throwing an exception in the `validate_options` method to cause a 500 internal server error, and the screen failed to account for the error scenario, causing the screen to go blank. I expected an error notice:

![image](https://user-images.githubusercontent.com/134745/87585928-9f66df00-c694-11ea-9f83-8a83b389f1d8.png)

johnwatkins0 committed:
Add unsaved changes warning to settings screen

johnwatkins0 committed:
Merge branch 'feature/5015-react-post-template-settings' of https://github.com/ampproject/amp-wp into feature/5015-react-post-template-settings

westonruter committed:
Ensure Content Types heading is shown in non-legacy reader mode

johnwatkins0 commented:
> @johnwatkins0 I just noticed the AYS dialog isn't showing up if I make a change and then try leaving the page. There should be some protected against a user accidentally leaving the page without saving changes, same as the wizard.

Added in 3a1888e72

westonruter committed:
Remove extraneous description for singular template

johnwatkins0 committed:
Add error boundary to settings screen

johnwatkins0 committed:
Merge branch 'feature/5015-react-post-template-settings' of https://github.com/ampproject/amp-wp into feature/5015-react-post-template-settings

westonruter committed:
Use confirm when unchecking the singular template

johnwatkins0 commented:
> Also, I tried throwing an exception in the `validate_options` method to cause a 500 internal server error, and the screen failed to account for the error scenario, causing the screen to go blank. I expected an error notice:
> 

Added the error boundary to the settings screen in dc06e114af8202e794316e3def48f4ac873c10d9

Pretty there will be some broken tests now for me to go over.

westonruter committed:
Add paragraph under Templates  with link to template hierarchy

westonruter commented:
> Added the error boundary to the settings screen in [dc06e11](https://github.com/ampproject/amp-wp/commit/dc06e114af8202e794316e3def48f4ac873c10d9)

That's better but I think we can do better. Namely, consider if a user's auth session expire or they log out in another tab. Currently this is the result:

![image](https://user-images.githubusercontent.com/134745/87589342-eefbd980-c699-11ea-8b49-304c5d73734e.png)

Their changes are lost. 

In spite of this, an auth will get offered by WordPress momentarily: 

![image](https://user-images.githubusercontent.com/134745/87589554-3d10dd00-c69a-11ea-8144-12ff16a62fd6.png)

Ideally instead of a full-screen error message, we'd show an error _notice_ instead. That will given the user a chance to recover. This is especially important when the user is in an offline state:

![image](https://user-images.githubusercontent.com/134745/87589683-75182000-c69a-11ea-8ed1-b13a2dc631e9.png)

johnwatkins0 commented:
Let me see what I can do real quick with a notice instead of a full-screen error.

johnwatkins0 commented:
@westonruter What do you think about this?

<img width="862" alt="Screen Shot 2020-07-15 at 4 14 52 PM" src="https://user-images.githubusercontent.com/9020968/87596826-60498580-c6b6-11ea-8e48-6fd1e99f0da4.png">

Once there is an error, this notice shows up and the save button is disabled until the page is refreshed. cc @jwold

westonruter commented:
@johnwatkins0 Yes, that's good. You could omit the “Refresh page“ link since in the case of the user being logged-out, the auth lightbox should soon reappear and in the case of being offline, refreshing won't help at all.

westonruter commented:
Oh, and make sure the Save Changes button becomes re-enabled so they can try again. The user needs to be able to try again:

![image](https://user-images.githubusercontent.com/134745/87607476-4443e500-c6b2-11ea-9dbd-cd4bc5aaa301.png)

westonruter committed:
Remove obsolete properties from supportable_templates response

westonruter committed:
Add readonly to supportable_post_types and supportable_templates

westonruter commented on assets/src/settings-page/settings-footer.js:
```diff
@@ -2,33 +2,43 @@
+		return () => undefined;
 	}, [ didSaveOptions, downloadingTheme ] );
 
 	const disabled = ! themeSupport ||
```
I just noticed that this needs to be updated to account for the initial state. 

When I first access the settings screen, I see:

![image](https://user-images.githubusercontent.com/134745/87601078-3342a600-c6a9-11ea-853c-7aa90a870715.png)

When I expected to see the disabled state:

![image](https://user-images.githubusercontent.com/134745/87601118-43f31c00-c6a9-11ea-834d-63babe925e80.png)

Aside: Let's make the gray darker because it has poor contrast for a11y.

westonruter replied:
Much better! 

![image](https://user-images.githubusercontent.com/134745/87607287-d8617c80-c6b1-11ea-99bc-35fe0d36b886.png)

johnwatkins0 committed:
Add error notice

johnwatkins0 committed:
Merge branch 'feature/5015-react-post-template-settings' of https://github.com/ampproject/amp-wp into feature/5015-react-post-template-settings

johnwatkins0 committed:
Simplify error handling

johnwatkins0 committed:
Disable footer button if there are no changes

johnwatkins0 committed:
Add more contrast to disabled button style

johnwatkins0 committed:
Rename useError to useAsyncError

johnwatkins0 committed:
Update disabled button bg color

johnwatkins0 committed:
Pass fullScreen prop to error boundary from wizard

westonruter committed:
Merge branch 'develop' of github.com:ampproject/amp-wp into feature/5015-react-post-template-settings

* 'develop' of github.com:ampproject/amp-wp: (34 commits)
  Add phpdoc for amp_server_timing_send action
  Remove unused index variable
  Adapt test for change in comma spacing
  Adapt output to reduce header size
  Add missing tests
  Only show verbose timings when logged in
  Fix ServerTiming hooks test
  Add missing throws annotation
  Add docblocks for server-timing actions
  Add tests for type assertion in Event
  Assert properties types
  Improve docblocks
  Fix copypasta docblock
  Fix bogus imports
  Use __METHOD__ instead of hard-coded string
  Remove empty files
  Add @covers tags for ServerTimingTest
  Move tests into correct namespace
  Complete ServerTiming tests
  Add beginning of ServerTimingTest
  ...

johnwatkins0 committed:
Fix linting error

johnwatkins0 committed:
Merge branch 'feature/5015-react-post-template-settings' of https://github.com/ampproject/amp-wp into feature/5015-react-post-template-settings

pierlon commented on assets/src/components/user-context-provider/index.js:
```diff
@@ -0,0 +1,142 @@
+ * Internal dependencies
+ */
+import { useAsyncError } from '../../utils/use-async-error';
+import { Options } from '../../components/options-context-provider';
```
```suggestion
import { Options } from '../options-context-provider';
```

johnwatkins0 commented:
@westonruter I have:

1. Implemented the error notice, although I think it has room for improvement. I need to step away from it and look again with fresh eyes later. Probably the wizard and the settings page just need separate error handlers. I don't think that needs to block merging, though, since we do have the basic error notice.
2. Adjusted the disabled button state
3. Disabled the button until there are changes

westonruter commented on assets/src/settings-page/settings-footer.js:
```diff
@@ -2,49 +2,70 @@
 
 	const disabled = ! themeSupport ||
+		! hasOptionsChanges ||
+		error ||
```
As noted in https://github.com/ampproject/amp-wp/pull/5044#issuecomment-659015797 I think the `error` condition needs to be removed here, or modified in some way so that if an error happened during saving the user can try saving again.

johnwatkins0 replied:
@westonruter I have the approach worked out and will be working through it.

johnwatkins0 replied:
I think 08e83151c4e96ddd28491779d899ad8fb601b766 will do what we need, but I need to double-check it this evening.

westonruter replied:
Yes, that seems to work well!

pierlon commented on assets/src/settings-page/supported-templates.js:
```diff
@@ -1,49 +1,243 @@
+				<li key={ supportableTemplate.id }>
+					<CheckboxControl
+						checked={ supportedTemplates.includes( supportableTemplate.id ) }
+						help={ supportableTemplate.description }
```
Seems `description` is not a property here:

![image](https://user-images.githubusercontent.com/16200219/87608848-75430a80-c6f0-11ea-90f1-287d32b8714a.png)

pierlon replied:
9105286f94624e36adf0687afd590c4aae25591c seems to be related.

westonruter replied:
It will be populated if it is defined. We removed the only `description` that was being used in `AMP_Theme_Support::get_supportable_templates()`, so that is why they are all empty. A plugin can use the `amp_supportable_templates` filter to add their own templates and provide a `description`.

pierlon replied:
Ah OK 👍.

westonruter committed:
Eliminate need for whitespace to be part of the translation string

pierlon commented on src/OptionsRESTController.php:
```diff
@@ -143,13 +158,59 @@ public function get_items( $request ) { // phpcs:ignore VariableAnalysis.CodeAna
-		$options[ self::SUPPRESSIBLE_PLUGINS ] = $this->plugin_suppression->get_suppressible_plugins_with_details();
+		$options[ self::SUPPRESSIBLE_PLUGINS ]   = $this->plugin_suppression->get_suppressible_plugins_with_details();
+		$options[ self::SUPPORTABLE_POST_TYPES ] = array_map(
+			function( $slug ) {
```
```suggestion
			static function( $slug ) {
```

johnwatkins0 committed:
Add separate error handling for errors not inside an error boundary

johnwatkins0 committed:
Merge branch 'feature/5015-react-post-template-settings' of https://github.com/ampproject/amp-wp into feature/5015-react-post-template-settings

westonruter committed:
Use static

Co-authored-by: Pierre Gordon <16200219+pierlon@users.noreply.github.com>

westonruter committed:
Simplify import path

Co-authored-by: Pierre Gordon <16200219+pierlon@users.noreply.github.com>

johnwatkins0 committed:
Update snapshot

westonruter committed:
Fix typo and improve variable naming

johnwatkins0 committed:
Remove snapshot test that wasn't useful

westonruter committed:
Merge branch 'develop' of github.com:ampproject/amp-wp into feature/5015-react-post-template-settings

* 'develop' of github.com:ampproject/amp-wp:
  Bump lodash from 4.17.15 to 4.17.19

johnwatkins0 committed:
Merge branch 'feature/5015-react-post-template-settings' of https://github.com/ampproject/amp-wp into feature/5015-react-post-template-settings

westonruter commented:
(I asked @pierlon to investigate the JS unit test failure introduced by merging #5052.)

johnwatkins0 committed:
Fill in error message if empty

pierlon committed:
Update package-lock.json

pierlon commented:
@googlebot I consent.

pierlon reviewed (approved):
Approving since the failing Optimizer tests are unrelated to the PR.

westonruter committed:
Merge branch 'develop' of github.com:ampproject/amp-wp into feature/5015-react-post-template-settings

* 'develop' of github.com:ampproject/amp-wp:
  Ignore write-only property errors
  Ignore unused private message errors

westonruter reviewed (approved):
😍
